File: src/claude/events.js
// ─── Tool metadata extraction ───
export function extractToolMeta(toolName, input) {
  const meta = { toolName, input: {} };
  if (input.file_path || input.path) meta.input.file = input.file_path || input.path;
  if (input.pattern) meta.input.pattern = input.pattern;
  if (input.command) meta.input.command = String(input.command).substring(0, 300);
  if (input.content) {
    meta.input.contentLength = input.content.length;
    meta.input.contentPreview = String(input.content).substring(0, 200);
  }
  if (input.old_string) {
    meta.input.editing = true;
    meta.input.oldString = String(input.old_string).substring(0, 150);
    meta.input.newString = String(input.new_string || '').substring(0, 150);
  }
  if (input.description) meta.input.description = String(input.description).substring(0, 200);
  if (input.url) meta.input.url = input.url;
  if (input.query) meta.input.query = String(input.query).substring(0, 200);
  if (input.prompt) meta.input.prompt = String(input.prompt).substring(0, 200);
  if (input.glob) meta.input.glob = input.glob;
  return meta;
}

function buildToolDisplay(toolName, meta) {
  let d = `Tool: ${toolName}`;
  if (meta.input.file) d += ` → ${meta.input.file}`;
  else if (meta.input.command) d += ` → ${meta.input.command}`;
  else if (meta.input.pattern) d += ` → ${meta.input.pattern}`;
  else if (meta.input.description) d += ` → ${meta.input.description}`;
  return d;
}

// ─── Main event handler ───
export function handleClaudeEvent(taskId, event, { addLog, queries, statsQueries, io, taskUsage, activeToolCalls }) {
  const type = event.type;

  switch (type) {
    case 'assistant': {
      const content = event.message?.content;
      if (Array.isArray(content)) {
        for (const block of content) {
          if (block.type === 'text' && block.text) {
            addLog(taskId, block.text, 'claude');
          } else if (block.type === 'tool_use') {
            const toolName = block.name || 'unknown';
            const toolId = block.id || null;
            const meta = extractToolMeta(toolName, block.input || {});
            meta.toolId = toolId;
            if (toolId) activeToolCalls.set(toolId, { taskId, toolName, startTime: Date.now(), meta });
            addLog(taskId, buildToolDisplay(toolName, meta), 'tool', meta);
          }
        }
      }

      // Live usage tracking
      const msgUsage = event.message?.usage;
      if (msgUsage) {
        const tracker = taskUsage.get(taskId);
        if (tracker) {
          tracker.session.input += msgUsage.input_tokens || 0;
          tracker.session.output += msgUsage.output_tokens || 0;
          tracker.session.cacheRead += msgUsage.cache_read_input_tokens || 0;
          tracker.session.cacheCreation += msgUsage.cache_creation_input_tokens || 0;

          const total = {
            input: tracker.baseline.input + tracker.session.input,
            output: tracker.baseline.output + tracker.session.output,
            cacheRead: tracker.baseline.cacheRead + tracker.session.cacheRead,
            cacheCreation: tracker.baseline.cacheCreation + tracker.session.cacheCreation,
            cost: tracker.baseline.cost,
          };

          const modelUsed = event.message?.model || '';
          queries.setTaskUsageLive.run(
            total.input,
            total.output,
            total.cacheRead,
            total.cacheCreation,
            total.cost,
            modelUsed,
            taskId,
          );
          io.emit('task:usage', {
            taskId,
            input_tokens: total.input,
            output_tokens: total.output,
            cache_read_tokens: total.cacheRead,
            cache_creation_tokens: total.cacheCreation,
            total_tokens: total.input + total.output,
            total_cost: total.cost,
          });
        }
      }
      break;
    }

    case 'user': {
      const content = event.message?.content;
      if (!Array.isArray(content)) break;
      for (const block of content) {
        if (block.type === 'tool_result') {
          const toolId = block.tool_use_id || null;
          const tracked = toolId ? activeToolCalls.get(toolId) : null;
          const duration = tracked ? Date.now() - tracked.startTime : null;
          const toolName = tracked?.toolName || 'unknown';

          let resultPreview = '';
          if (typeof block.content === 'string') resultPreview = block.content.substring(0, 500);
          else if (Array.isArray(block.content)) {
            resultPreview = block.content
              .filter((c) => c.type === 'text')
              .map((c) => c.text)
              .join('\n')
              .substring(0, 500);
          }
          const resultLines = resultPreview.split('\n').length;
          const isError = block.is_error === true;

          const meta = { toolName, toolId, duration, isResult: true, isError, resultPreview, resultLines };
          const durationStr = duration ? ` (${duration}ms)` : '';
          const icon = isError ? '✗' : '✓';
          let display = `${icon} ${toolName}${durationStr}`;
          if (resultPreview) {
            const first = resultPreview.split('\n')[0].substring(0, 120);
            if (first.trim()) display += ` — ${first}`;
          }
          addLog(taskId, display, isError ? 'error' : 'tool_result', meta);
          if (toolId) activeToolCalls.delete(toolId);
        }
      }
      break;
    }

    case 'result': {
      const usage = event.usage || {};
      const sessionInput = usage.input_tokens || 0;
      const sessionOutput = usage.output_tokens || 0;
      const sessionCacheRead = usage.cache_read_input_tokens || 0;
      const sessionCacheCreation = usage.cache_creation_input_tokens || 0;
      const totalCost = event.total_cost || 0;
      const numTurns = event.num_turns || 0;
      const durationMs = event.duration_ms || 0;
      const modelUsed = event.model || '';
      const sessionId = event.session_id || '';

      const tracker = taskUsage.get(taskId);
      const base = tracker?.baseline || { input: 0, output: 0, cacheRead: 0, cacheCreation: 0, cost: 0 };

      const fin = {
        input: base.input + sessionInput,
        output: base.output + sessionOutput,
        cacheRead: base.cacheRead + sessionCacheRead,
        cacheCreation: base.cacheCreation + sessionCacheCreation,
        cost: base.cost + totalCost,
      };

      if (sessionInput > 0 || sessionOutput > 0) {
        queries.setTaskUsageLive.run(
          fin.input,
          fin.output,
          fin.cacheRead,
          fin.cacheCreation,
          fin.cost,
          modelUsed,
          taskId,
        );
        if (numTurns > 0) queries.updateTaskNumTurns.run(numTurns, taskId);
        if (sessionId) queries.updateTaskClaudeSession.run(sessionId, taskId);

        const costStr = totalCost > 0 ? ` | Cost: $${totalCost.toFixed(4)}` : '';
        const durStr = durationMs > 0 ? ` | Duration: ${Math.round(durationMs / 1000)}s` : '';
        addLog(
          taskId,
          `Usage: ${(sessionInput + sessionOutput).toLocaleString()} tokens (${sessionInput.toLocaleString()} in / ${sessionOutput.toLocaleString()} out)${costStr} | Turns: ${numTurns}${durStr} | Model: ${modelUsed}`,
          'system',
        );

        const updated = queries.getTaskById.get(taskId);
        if (updated) {
          io.emit('task:updated', updated);
          io.emit('task:usage', {
            taskId,
            input_tokens: fin.input,
            output_tokens: fin.output,
            cache_read_tokens: fin.cacheRead,
            cache_creation_tokens: fin.cacheCreation,
            total_tokens: fin.input + fin.output,
            total_cost: fin.cost,
          });
        }
      }

      // Save model info to limits table
      const modelUsage = event.modelUsage;
      if (modelUsage) {
        const firstModel = Object.keys(modelUsage)[0];
        const mu = modelUsage[firstModel] || {};
        try {
          statsQueries.upsertClaudeLimits({
            rateLimitType: '',
            status: 'allowed',
            resetsAt: 0,
            overageStatus: '',
            isUsingOverage: false,
            model: firstModel || modelUsed,
            costUsd: event.total_cost_usd || totalCost || 0,
            contextWindow: mu.contextWindow || 0,
            maxOutputTokens: mu.maxOutputTokens || 0,
          });
          io.emit('claude:limits', {
            model: firstModel,
            costUsd: event.total_cost_usd || totalCost,
            contextWindow: mu.contextWindow,
            maxOutputTokens: mu.maxOutputTokens,
          });
        } catch {}
      }

      if (event.result) addLog(taskId, `Result: ${String(event.result).substring(0, 500)}`, 'success');
      break;
    }

    case 'system': {
      const subtype = event.subtype || '';
      if (subtype === 'hook_started' || subtype === 'hook_response') break;
      if (subtype === 'init') {
        addLog(taskId, `Session initialized (${event.tools?.length || 0} tools available)`, 'system');
        break;
      }
      const msg = event.message || '';
      if (
        msg.toLowerCase().includes('rate limit') ||
        msg.toLowerCase().includes('429') ||
        msg.toLowerCase().includes('overloaded')
      ) {
        queries.incrementRateLimitHits.run(taskId);
      }
      if (msg) addLog(taskId, msg, 'system');
      break;
    }

    case 'rate_limit_event': {
      const info = event.rate_limit_info || {};

      // Always save latest limit info to DB
      try {
        statsQueries.upsertClaudeLimits({
          rateLimitType: info.rateLimitType || '',
          status: info.status || '',
          resetsAt: info.resetsAt || 0,
          overageStatus: info.overageStatus || '',
          isUsingOverage: info.isUsingOverage || false,
          model: '',
          costUsd: 0,
          contextWindow: 0,
          maxOutputTokens: 0,
        });
        io.emit('claude:limits', {
          rateLimitType: info.rateLimitType,
          status: info.status,
          resetsAt: info.resetsAt,
          overageStatus: info.overageStatus,
          isUsingOverage: info.isUsingOverage,
        });
      } catch {}

      if (info.status !== 'allowed') {
        queries.incrementRateLimitHits.run(taskId);
        const resetAt = info.resetsAt ? new Date(info.resetsAt * 1000).toLocaleTimeString() : '';
        addLog(
          taskId,
          `Rate limited (${info.rateLimitType || 'unknown'})${resetAt ? ` — resets at ${resetAt}` : ''}`,
          'error',
        );
      }
      break;
    }
  }
}

File: src-tauri/src/claude/events.rs
use serde_json::Value;
use tauri::{AppHandle, Emitter};
use crate::db::{self, DbPool};
use crate::db::tasks;
use crate::db::stats;
use std::collections::HashMap;
use std::sync::Mutex;

pub struct UsageTracker {
    pub baseline: UsageBaseline,
    pub session: UsageSession,
}

pub struct UsageBaseline {
    pub input: i64,
    pub output: i64,
    pub cache_read: i64,
    pub cache_creation: i64,
    pub cost: f64,
}

#[derive(Default)]
pub struct UsageSession {
    pub input: i64,
    pub output: i64,
    pub cache_read: i64,
    pub cache_creation: i64,
}

pub struct ToolCall {
    pub task_id: i64,
    pub tool_name: String,
    pub start_time: std::time::Instant,
}

pub struct EventContext {
    pub task_usage: Mutex<HashMap<i64, UsageTracker>>,
    pub active_tool_calls: Mutex<HashMap<String, ToolCall>>,
}

impl EventContext {
    pub fn new() -> Self {
        Self {
            task_usage: Mutex::new(HashMap::new()),
            active_tool_calls: Mutex::new(HashMap::new()),
        }
    }
}

pub fn handle_event(
    task_id: i64,
    event: &Value,
    db: &DbPool,
    app: &AppHandle,
    ctx: &EventContext,
) {
    let event_type = event.get("type").and_then(|v| v.as_str()).unwrap_or("");

    match event_type {
        "assistant" => handle_assistant(task_id, event, db, app, ctx),
        "user" => handle_user(task_id, event, db, app, ctx),
        "result" => handle_result(task_id, event, db, app, ctx),
        "system" => handle_system(task_id, event, db, app),
        "rate_limit_event" => handle_rate_limit(task_id, event, db, app),
        _ => {}
    }
}

fn add_log(task_id: i64, message: &str, log_type: &str, db: &DbPool, app: &AppHandle, meta: Option<&str>) {
    tasks::add_log(db, task_id, message, log_type, meta);
    let payload = serde_json::json!({
        "taskId": task_id,
        "message": message,
        "logType": log_type,
        "created_at": chrono::Local::now().to_rfc3339(),
    });
    app.emit("task:log", &payload).ok();
}

fn handle_assistant(task_id: i64, event: &Value, db: &DbPool, app: &AppHandle, ctx: &EventContext) {
    let content = event.pointer("/message/content");
    if let Some(blocks) = content.and_then(|c| c.as_array()) {
        for block in blocks {
            let block_type = block.get("type").and_then(|v| v.as_str()).unwrap_or("");
            match block_type {
                "text" => {
                    if let Some(text) = block.get("text").and_then(|v| v.as_str()) {
                        add_log(task_id, text, "claude", db, app, None);
                    }
                }
                "tool_use" => {
                    let tool_name = block.get("name").and_then(|v| v.as_str()).unwrap_or("unknown");
                    let tool_id = block.get("id").and_then(|v| v.as_str()).unwrap_or("");
                    let input = block.get("input").cloned().unwrap_or(Value::Object(Default::default()));

                    let mut display = format!("Tool: {}", tool_name);
                    if let Some(f) = input.get("file_path").or(input.get("path")).and_then(|v| v.as_str()) {
                        display = format!("{} → {}", display, f);
                    } else if let Some(c) = input.get("command").and_then(|v| v.as_str()) {
                        display = format!("{} → {}", display, &c[..c.len().min(120)]);
                    } else if let Some(p) = input.get("pattern").and_then(|v| v.as_str()) {
                        display = format!("{} → {}", display, p);
                    }

                    let meta = serde_json::json!({"toolName": tool_name, "toolId": tool_id}).to_string();
                    add_log(task_id, &display, "tool", db, app, Some(&meta));

                    if !tool_id.is_empty() {
                        ctx.active_tool_calls.lock().unwrap().insert(tool_id.to_string(), ToolCall {
                            task_id,
                            tool_name: tool_name.to_string(),
                            start_time: std::time::Instant::now(),
                        });
                    }
                }
                _ => {}
            }
        }
    }

    // Usage tracking
    if let Some(usage) = event.pointer("/message/usage") {
        let mut trackers = ctx.task_usage.lock().unwrap();
        if let Some(tracker) = trackers.get_mut(&task_id) {
            tracker.session.input += usage.get("input_tokens").and_then(|v| v.as_i64()).unwrap_or(0);
            tracker.session.output += usage.get("output_tokens").and_then(|v| v.as_i64()).unwrap_or(0);
            tracker.session.cache_read += usage.get("cache_read_input_tokens").and_then(|v| v.as_i64()).unwrap_or(0);
            tracker.session.cache_creation += usage.get("cache_creation_input_tokens").and_then(|v| v.as_i64()).unwrap_or(0);

            let total_input = tracker.baseline.input + tracker.session.input;
            let total_output = tracker.baseline.output + tracker.session.output;
            let total_cr = tracker.baseline.cache_read + tracker.session.cache_read;
            let total_cc = tracker.baseline.cache_creation + tracker.session.cache_creation;
            let total_cost = tracker.baseline.cost;

            let model_used = event.pointer("/message/model").and_then(|v| v.as_str()).unwrap_or("");
            tasks::set_usage_live(db, task_id, total_input, total_output, total_cr, total_cc, total_cost, model_used);

            app.emit("task:usage", &serde_json::json!({
                "taskId": task_id,
                "input_tokens": total_input,
                "output_tokens": total_output,
                "cache_read_tokens": total_cr,
                "cache_creation_tokens": total_cc,
                "total_tokens": total_input + total_output,
                "total_cost": total_cost,
            })).ok();
        }
    }
}

fn handle_user(task_id: i64, event: &Value, db: &DbPool, app: &AppHandle, ctx: &EventContext) {
    let content = event.pointer("/message/content");
    if let Some(blocks) = content.and_then(|c| c.as_array()) {
        for block in blocks {
            if block.get("type").and_then(|v| v.as_str()) == Some("tool_result") {
                let tool_id = block.get("tool_use_id").and_then(|v| v.as_str()).unwrap_or("");
                let tracked = ctx.active_tool_calls.lock().unwrap().remove(tool_id);
                let duration = tracked.as_ref().map(|t| t.start_time.elapsed().as_millis() as i64);
                let tool_name = tracked.as_ref().map(|t| t.tool_name.as_str()).unwrap_or("unknown");
                let is_error = block.get("is_error").and_then(|v| v.as_bool()).unwrap_or(false);

                let result_preview = if let Some(s) = block.get("content").and_then(|v| v.as_str()) {
                    s[..s.len().min(500)].to_string()
                } else {
                    String::new()
                };

                let icon = if is_error { "✗" } else { "✓" };
                let dur_str = duration.map(|d| format!(" ({}ms)", d)).unwrap_or_default();
                let first_line = result_preview.lines().next().unwrap_or("").chars().take(120).collect::<String>();
                let display = if first_line.trim().is_empty() {
                    format!("{} {}{}", icon, tool_name, dur_str)
                } else {
                    format!("{} {}{} — {}", icon, tool_name, dur_str, first_line)
                };

                let lt = if is_error { "error" } else { "tool_result" };
                add_log(task_id, &display, lt, db, app, None);
            }
        }
    }
}

fn handle_result(task_id: i64, event: &Value, db: &DbPool, app: &AppHandle, ctx: &EventContext) {
    let usage = event.get("usage").cloned().unwrap_or(Value::Object(Default::default()));
    let session_input = usage.get("input_tokens").and_then(|v| v.as_i64()).unwrap_or(0);
    let session_output = usage.get("output_tokens").and_then(|v| v.as_i64()).unwrap_or(0);
    let session_cr = usage.get("cache_read_input_tokens").and_then(|v| v.as_i64()).unwrap_or(0);
    let session_cc = usage.get("cache_creation_input_tokens").and_then(|v| v.as_i64()).unwrap_or(0);
    let total_cost = event.get("total_cost").and_then(|v| v.as_f64()).unwrap_or(0.0);
    let num_turns = event.get("num_turns").and_then(|v| v.as_i64()).unwrap_or(0);
    let duration_ms = event.get("duration_ms").and_then(|v| v.as_i64()).unwrap_or(0);
    let model_used = event.get("model").and_then(|v| v.as_str()).unwrap_or("");
    let session_id = event.get("session_id").and_then(|v| v.as_str()).unwrap_or("");

    let trackers = ctx.task_usage.lock().unwrap();
    let base = trackers.get(&task_id).map(|t| &t.baseline);
    let bi = base.map(|b| b.input).unwrap_or(0);
    let bo = base.map(|b| b.output).unwrap_or(0);
    let bcr = base.map(|b| b.cache_read).unwrap_or(0);
    let bcc = base.map(|b| b.cache_creation).unwrap_or(0);
    let bc = base.map(|b| b.cost).unwrap_or(0.0);
    drop(trackers);

    let fin_input = bi + session_input;
    let fin_output = bo + session_output;
    let fin_cr = bcr + session_cr;
    let fin_cc = bcc + session_cc;
    let fin_cost = bc + total_cost;

    if session_input > 0 || session_output > 0 {
        tasks::set_usage_live(db, task_id, fin_input, fin_output, fin_cr, fin_cc, fin_cost, model_used);
        if num_turns > 0 { tasks::update_num_turns(db, task_id, num_turns); }
        if !session_id.is_empty() { tasks::update_claude_session(db, task_id, session_id); }

        let cost_str = if total_cost > 0.0 { format!(" | Cost: ${:.4}", total_cost) } else { String::new() };
        let dur_str = if duration_ms > 0 { format!(" | Duration: {}s", duration_ms / 1000) } else { String::new() };
        let msg = format!(
            "Usage: {} tokens ({} in / {} out){} | Turns: {}{} | Model: {}",
            session_input + session_output, session_input, session_output, cost_str, num_turns, dur_str, model_used
        );
        add_log(task_id, &msg, "system", db, app, None);

        if let Some(updated) = tasks::get_by_id(db, task_id) {
            app.emit("task:updated", &updated).ok();
            app.emit("task:usage", &serde_json::json!({
                "taskId": task_id,
                "input_tokens": fin_input, "output_tokens": fin_output,
                "cache_read_tokens": fin_cr, "cache_creation_tokens": fin_cc,
                "total_tokens": fin_input + fin_output, "total_cost": fin_cost,
            })).ok();
        }
    }

    // Save model limits
    if let Some(model_usage) = event.get("modelUsage").and_then(|v| v.as_object()) {
        if let Some((first_model, mu)) = model_usage.iter().next() {
            stats::upsert_claude_limits(
                &db::get_db(), "", "allowed", 0, "", false,
                first_model, event.get("total_cost_usd").and_then(|v| v.as_f64()).unwrap_or(total_cost),
                mu.get("contextWindow").and_then(|v| v.as_i64()).unwrap_or(0),
                mu.get("maxOutputTokens").and_then(|v| v.as_i64()).unwrap_or(0),
            );
        }
    }

    if let Some(result) = event.get("result").and_then(|v| v.as_str()) {
        let preview = &result[..result.len().min(500)];
        add_log(task_id, &format!("Result: {}", preview), "success", db, app, None);
    }
}

fn handle_system(task_id: i64, event: &Value, db: &DbPool, app: &AppHandle) {
    let subtype = event.get("subtype").and_then(|v| v.as_str()).unwrap_or("");
    if subtype == "hook_started" || subtype == "hook_response" { return; }
    if subtype == "init" {
        let tools = event.get("tools").and_then(|v| v.as_array()).map(|a| a.len()).unwrap_or(0);
        add_log(task_id, &format!("Session initialized ({} tools available)", tools), "system", db, app, None);
        return;
    }
    let msg = event.get("message").and_then(|v| v.as_str()).unwrap_or("");
    if !msg.is_empty() {
        let lower = msg.to_lowercase();
        if lower.contains("rate limit") || lower.contains("429") || lower.contains("overloaded") {
            tasks::increment_rate_limit_hits(db, task_id);
        }
        add_log(task_id, msg, "system", db, app, None);
    }
}

fn handle_rate_limit(task_id: i64, event: &Value, db: &DbPool, app: &AppHandle) {
    let info = event.get("rate_limit_info").cloned().unwrap_or(Value::Object(Default::default()));
    let rlt = info.get("rateLimitType").and_then(|v| v.as_str()).unwrap_or("");
    let status = info.get("status").and_then(|v| v.as_str()).unwrap_or("");
    let resets_at = info.get("resetsAt").and_then(|v| v.as_i64()).unwrap_or(0);
    let overage_status = info.get("overageStatus").and_then(|v| v.as_str()).unwrap_or("");
    let is_using_overage = info.get("isUsingOverage").and_then(|v| v.as_bool()).unwrap_or(false);

    stats::upsert_claude_limits(
        &db::get_db(), rlt, status, resets_at,
        overage_status, is_using_overage,
        "", 0.0, 0, 0,
    );

    app.emit("claude:limits", &serde_json::json!({
        "rateLimitType": rlt,
        "status": status,
        "resets_at": resets_at,
        "overageStatus": overage_status,
        "isUsingOverage": is_using_overage,
    })).ok();

    if status != "allowed" {
        tasks::increment_rate_limit_hits(db, task_id);
        let msg = format!("Rate limited ({})", rlt);
        add_log(task_id, &msg, "error", db, app, None);
    }
}
